lib/libc/symtab/symtab_findorderedbyname.c

```diff
@@ -0,0 +1,134 @@
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+
+#include <string.h>
+#include <debug.h>
+#include <assert.h>
+#include <errno.h>
+
+#include <tinyara/symtab.h>
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: symtab_findorderedbyname
+ *
+ * Description:
+ *   Find the symbol in the symbol table with the matching name.
+ *   This version assumes that table ordered with respect to symbol name.
+ *
+ *   This function is a lot larger than symbtab_findbyname().  This function
+ *   not be used, unless the symbol table is large and the performance
+ *   benefit is worth the increased size.
+ *
+ * Returned Value:
+ *   A reference to the symbol table entry if an entry with the matching
+ *   name is found; NULL is returned if the entry is not found.
+ *
+ ****************************************************************************/
+
+FAR const struct symtab_s *symtab_findorderedbyname(FAR const struct symtab_s *symtab, FAR const char *name, int nsyms)
+{
+	int low = 0;
+	int high = nsyms - 1;
+	int mid;
+	int cmp;
+
+	/* Loop until the range has been isolated to a single symbol table
+	 * entry that may or may not match the search name.
+	 */
+
+	DEBUGASSERT(symtab != NULL && name != NULL);
+	while (low < high) {
+		/* Compare the name to the one in the middle.  (or just below
+		 * the middle in the case where one is even and one is odd).
+		 */
+
+		mid = (low + high) >> 1;
+		cmp = strcmp(name, symtab[mid].sym_name);
+		if (cmp < 0) {
+			/* name < symtab[mid].sym_name
+			 *
+			 * NOTE: Because of truncation in the calculation of 'mid'.
+			 * 'mid' could be equal to 'low'
+			 */
+
+			high = mid > low ? mid - 1 : low;
```

Comment:
I think no need to equate high to low. Because in such a case, high will be equal to low -1.